### PR TITLE
Torch baselines for attention benchmarks

### PR DESCRIPTION
**Pull request by Priya2698**
Issue #1668

The benchmarks use `dropout_p=0.2` for benchmarking. For validating in forward pass, to ensure the same dropout_mask is applied, this is set to 0.

### CODE REVIEW

**Review by naoyam — APPROVED** [review]
LGTM

File: benchmarks/python/test_huggingface_attn_bwd.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import pytest
 from nvfuser import FusionDefinition, DataType
 from nvfuser.pytorch_utils import torch_dtype_to_nvfuser_dtype
-from .core import run_benchmark, clear_cuda_cache
+from .core import run_benchmark, clear_cuda_cache, unary_bwd_torch
```
Comment (liqiangxl) [inline benchmarks/python/test_huggingface_attn_bwd.py:4]: where does this `unary_bwd_torch` defined? It doesn't exist in `core.py`.

File: benchmarks/python/test_huggingface_attn_bwd.py [hunk 2]
```diff
@@ -95,3 +101,34 @@ def test_huggingface_attn_bwd_benchmark(
...
+    attn = torch.nn.functional.softmax(attn, dim=-1)
+    output = torch.nn.functional.dropout(attn, p=dropout_p)
+
+    grads = torch.randn(batch_size * nh, seq_len, seq_len, device="cuda", dtype=dtype)
+
+    run_benchmark(
+        benchmark,
+        torch.compile(unary_bwd_torch) if compile else unary_bwd_torch,
+        [output, grads],
+        iobytes=huggingface_attn_bwd_iobytes(size, dtype),
```
Comment (liqiangxl) [inline benchmarks/python/test_huggingface_attn_bwd.py:133]: can you comment why we need to set `iobytes`? what's the difference is using default.

File: benchmarks/python/test_huggingface_attn_fwd.py [hunk 3]
```diff
@@ -71,9 +71,26 @@ def huggingface_attn_fwd_fusion(
     fd.add_output(T47)
 
 
+def huggingface_attn_fwd_fn(
```
Comment (liqiangxl) [inline benchmarks/python/test_huggingface_attn_fwd.py:74]: do we have references for this and all the other baseline implementations?

File: benchmarks/python/test_huggingface_attn_fwd.py [hunk 4]
```diff
@@ -103,3 +120,28 @@ def test_huggingface_attn_fwd_benchmark(
...
+def test_huggingface_attn_fwd_baseline_benchmark(
+    benchmark,
+    size: tuple,
+    dtype: torch.dtype,
+    compile: bool,
+):
+    clear_cuda_cache()
+
+    batch_size, seq_len, nh, n_embd = size
+    dropout_p = 0.0
```
Comment (liqiangxl) [inline benchmarks/python/test_huggingface_attn_fwd.py:137]: is there an influence on perf if dropout is larger than 0, e.g. 0.2? If has, we can use a more practical value and set to 0 if validation is turned on.

Comment (liqiangxl) [reply benchmarks/python/test_huggingface_attn_fwd.py:137]: If its `0`, the predication is trivial and may influence the perf.
```
   S9 = fd.define_scalar(1 - dropout_p, dtype=DataType.Double)
    T10 = fd.ops.lt(T8, S9)
```

File: benchmarks/python/test_huggingface_attn_bwd.py [hunk 5]
```diff
@@ -55,9 +55,15 @@ def huggingface_attn_bwd_fusion(
     fd.add_output(T35)
 
 
+def huggingface_attn_bwd_iobytes(size: tuple, dtype: torch.dtype):
```
Comment (jacobhinkle) [inline benchmarks/python/test_huggingface_attn_bwd.py:58]: tiny nit: when annotating collections, specify the element types. For tuple, annotate all of them like `size: tuple[int, int, int, int]`.

Comment (jacobhinkle) [reply benchmarks/python/test_huggingface_attn_bwd.py:58]: As @liqiangxl already mentioned, if this just implements the default behavior i.e. `sum(i.nbytes for i in inputs) + sum(o.nbytes for o in outputs)` then we should remove this helper.

File: benchmarks/python/test_huggingface_attn_bwd.py [hunk 6]
```diff
@@ -95,3 +101,34 @@ def test_huggingface_attn_bwd_benchmark(
...
     if not disable_benchmarking:
         run_benchmark(benchmark, fd.execute, [grads, attn, dropout_mask])
+
+
+@pytest.mark.parametrize("compile", [False, True], ids=["eager", "compile"])
+@pytest.mark.parametrize("size", generate_attn_inputs())
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
+def test_huggingface_attn_bwd_baseline_benchmark(
+    benchmark,
+    size: tuple,
```
Comment (jacobhinkle) [inline benchmarks/python/test_huggingface_attn_bwd.py:111]: Nit: `size: tuple[int, int, int, int]`

File: benchmarks/python/test_huggingface_attn_fwd.py [hunk 7]
```diff
@@ -71,9 +71,26 @@ def huggingface_attn_fwd_fusion(
     fd.add_output(T47)
 
 
+def huggingface_attn_fwd_fn(
+    inputs: list,
+):  # [attention_mask, in_tensor, size, dropout_p]
```
Comment (jacobhinkle) [inline benchmarks/python/test_huggingface_attn_fwd.py:76]: Comment should be a docstring if providing more details, or omitted since it's clear from the following line.

File: benchmarks/python/test_huggingface_attn_fwd.py (hunk 3 above)
Comment (jacobhinkle) [inline benchmarks/python/test_huggingface_attn_fwd.py:74]: Might I suggest
```suggestion
def huggingface_attn_fwd(
```
And related changes further down in this file.

File: benchmarks/python/test_huggingface_attn_bwd.py (hunk 1 above)
Comment (Priya2698) [reply benchmarks/python/test_huggingface_attn_bwd.py:4]: It is defined here: https://github.com/NVIDIA/Fuser/blob/e10c4714aede47a41a1d71a879681d3f94c00df0/benchmarks/python/core.py#L142

File: benchmarks/python/test_huggingface_attn_bwd.py (hunk 2 above)
Comment (Priya2698) [reply benchmarks/python/test_huggingface_attn_bwd.py:133]: This is for uniformity of IObytes between nvFuser and torch when the inputs/outputs differ between the two. See the discussion at: https://github.com/NVIDIA/Fuser/pull/1725. I'll add a reference to this PR in a comment.

File: benchmarks/python/test_huggingface_attn_fwd.py (hunk 3 above)
Comment (Priya2698) [reply benchmarks/python/test_huggingface_attn_fwd.py:74]: I'll add links to the thunder code: They were removed earlier. 
Is that what you're referring to?

File: benchmarks/python/test_huggingface_attn_fwd.py (hunk 4 above)
Comment (Priya2698) [reply benchmarks/python/test_huggingface_attn_fwd.py:137]: Right. Let me make that change. Thanks!

File: benchmarks/python/test_huggingface_attn_bwd.py (hunk 5 above)
Comment (Priya2698) [reply benchmarks/python/test_huggingface_attn_bwd.py:58]: This is for uniformity of IObytes between nvFuser and torch when the inputs/outputs differ between the two. See the discussion at: https://github.com/NVIDIA/Fuser/pull/1725. I'll add a reference to this PR in a comment.

Comment (Priya2698) [reply benchmarks/python/test_huggingface_attn_bwd.py:58]: > tiny nit: when annotating collections, specify the element types. For tuple, annotate all of them like `size: tuple[int, int, int, int]`.

I think for size, the element types are understood. So I prefer this to reduce verbosity. Not a strong preference, though. I may have a bias being more familiar with the benchmarks.

Comment (jacobhinkle) [reply benchmarks/python/test_huggingface_attn_bwd.py:58]: Does the IOBytes issue affect this computation though? It seems like the helper method is just replicating default behavior. If that's the case, then maybe putting argument `iobytes=None,` would be the clearest. As you wish on the tuple annotation.

Comment (Priya2698) [reply benchmarks/python/test_huggingface_attn_bwd.py:58]: The nvFuser input/outputs are grad_out, attention_mask, dropout_mask and grad_input.
The eager input/outputs are grad_out and output. So there will be a difference in the computed IObytes.

Comment (jacobhinkle) [reply benchmarks/python/test_huggingface_attn_bwd.py:58]: Oh ok, so it does affect this case due to recomputation. Could you place a comment in the code describing that? Thanks.

File: benchmarks/python/test_huggingface_attn_fwd.py (hunk 4 above)
Comment (Priya2698) [reply benchmarks/python/test_huggingface_attn_fwd.py]: This would although mean that we benchmark different things based on whether validation is turned on.

Comment (liqiangxl) [reply benchmarks/python/test_huggingface_attn_fwd.py]: can you check the influence on bandwidth? I am not sure how much it may have, maybe very small.

Comment (Priya2698) [reply benchmarks/python/test_huggingface_attn_fwd.py]: What do you think about having another fusion definition with 0.0 dropout in validation and 0.2 otherwise. That would ensure we always benchmark with dropout (0.2) and it is consistent with torch baseline benchmarks.

Comment (Priya2698) [reply benchmarks/python/test_huggingface_attn_fwd.py]: Dropout: 0.2
Screenshot 2024-04-04 at 11 53 57 AM

Dropout: 0.0
Screenshot 2024-04-04 at 11 53 49 AM

The difference for eager is significant.

Comment (liqiangxl) [reply benchmarks/python/test_huggingface_attn_fwd.py:140]: > What do you think about having another fusion definition with 0.0 dropout in validation and 0.2 otherwise. That would ensure we always benchmark with dropout (0.2) and it is consistent with torch baseline benchmarks.

sounds good to me.

Comment (liqiangxl) [reply benchmarks/python/test_huggingface_attn_fwd.py:140]: > Dropout: 0.2 Screenshot 2024-04-04 at 11 53 57 AM
> 
> Dropout: 0.0 Screenshot 2024-04-04 at 11 53 49 AM
> 
> The difference for eager is significant.

This seems difficult to explain. can you create an issue to investigate further when we have time?

Comment (jacobhinkle) [reply benchmarks/python/test_huggingface_attn_fwd.py:140]: Eager mode is probably skipping the dropout call altogether when `dropout=0` but i would guess torch.compile is either not skipping dropout, or it's just io-bound so skipping dropout makes no difference.

Comment (liqiangxl) [reply benchmarks/python/test_huggingface_attn_fwd.py:140]: that's very smart!

**Comment by Priya2698** [comment]
Updated the PR description.
We use a non-zero dropout in benchmarking now. For fwd calls, I set it to 0 for validation.

Discovered a bug when using non-zero dropout in nanogpt bwd pass that is now fixed.
